runtime/commonMain/src/kotlinx/serialization/Annotations.kt

```diff
@@ -10,60 +10,156 @@ import kotlinx.serialization.json.*
+ * class RgbPixel(val red: Byte, val green: Byte, val blue: Byte)
+ *
+ * @Serializable
+ * class RgbExample(
+ *     @Serializable(with = RgbAsHexString::class) p1: RgpPixel, // Serialize as HEX string, e.g. #FFFF00
+ *     @Serializable(with = RgbAsSingleInt::class) p2: RgpPixel, // Serialize as single integer, e.g. 16711680
+ *     p3: RgpPixel // Serialize as 3 byte components, e.g. { "red": 255, "green": 255, "blue": 0 }
+ * )
+ * ```
+ * In this example, each pixel will be serialized using different data representation.
+ *
+ * For classes with generic type parameters, `serializer()` method requires one additional argument per each generic type parameter:
+ * ```
+ * @Serializable
+ * class Box<T>(value: T)
+ *
+ * Box.serializer() // Doesn't compile
+ * Box.serializer(Int.serializer()) // Returns serializer for Box<Int>
+ * Box.serializer(Box.serializer(Int.serializer())) // Returns serializer for Box<Box<Int>>
+ * ```
+ *
+ * ### Implementation details
+ *
+ * In order to generate `serializer` method that is not a method on the particular instance, the class should have a companion object, either named or unnamed.
+ * Companion object is generated by the plugin if it is not declared, effectively exposing both companion and `serializer()` method to class ABI.
+ * If companion object already exists, only `serializer` method will be generated.
+ * If `serializer` method already exists, it will be matched against the desired signature `serializer(vararg type params): KSerializer<T>`.
+ * On successful match, the existing method will be used as [KSerializer] source instead of auto-generated one.
+ * On unsuccessful match, a compiler error occurs.
  *
  * @see UseSerializers
+ * @see Serializer
  */
 @Target(AnnotationTarget.PROPERTY, AnnotationTarget.CLASS, AnnotationTarget.TYPE)
+@Retention(AnnotationRetention.BINARY)
 public annotation class Serializable(
-    val with: KClass<out KSerializer<*>> = KSerializer::class // it means -- use default serializer by default
+    val with: KClass<out KSerializer<*>> = KSerializer::class // Default value indicates that auto-generated serializer is used
 )
 
 /**
- * Instructs plugin to turn this class into serializer for specified class [forClass].
+ * Instructs the serialization plugin to turn this class into serializer for specified class [forClass].
  * However, it would not be used automatically. To apply it on particular class or property,
  * use [Serializable] or [UseSerializers], or [ContextualSerialization] with runtime registration.
  */
 @Target(AnnotationTarget.CLASS)
+@Retention(AnnotationRetention.BINARY)
 public annotation class Serializer(
-    val forClass: KClass<*> // what class to create serializer for
+    val forClass: KClass<*> //  target class to create serializer for
 )
 
 /**
- * Overrides name visible to the runtime part of serialization framework
+ * Overrides the name of the current element in the [SerialDescriptor].
```

Comment:
It's not clear what 'current element' refers to (unless one carefully reads documentation to Encoder and SerialDescriptor). Better simply use 'class or property'